sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -98,6 +107,138 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
+      PhysicalOperation(
+        rightProjections,
+        Nil,
+        rightHolder @ ScanBuilderHolder(_, _, rBuilder: SupportsPushDownJoin)
+      ),
+      joinType,
+      condition,
+    _) if conf.dataSourceV2JoinPushdown &&
+      // We do not support pushing down anything besides AttributeReference.
+      leftProjections.forall(_.isInstanceOf[AttributeReference]) &&
+      rightProjections.forall(_.isInstanceOf[AttributeReference]) &&
+      // Cross joins are not supported because they increase the amount of data.
+      condition.isDefined &&
+      // Joins on top of sampled tables are not supported
+      leftHolder.pushedSample.isEmpty &&
+      rightHolder.pushedSample.isEmpty &&
+      lBuilder.isOtherSideCompatibleForJoin(rBuilder) =>
+
+      // projections' names are maybe not up to date if the joins have been previously pushed down.
+      // For this reason, we need to use pushedJoinOutputMap to get up to date names.
+      //
+      // Normalized projections are then converted to StructType.
+      def getRequiredSchema(
+          projections: Seq[NamedExpression],
+          sHolder: ScanBuilderHolder): StructType = {
+        val normalizedProjections = DataSourceStrategy.normalizeExprs(
+          projections,
+          sHolder.output.map { a =>
+            sHolder.pushedJoinOutputMap.getOrElse(a, a).asInstanceOf[AttributeReference]
+          }
+        ).asInstanceOf[Seq[AttributeReference]]
+
+        fromAttributes(normalizedProjections)
+      }
+
+      def generateJoinOutputAlias(name: String): String =
+        s"${name}_${java.util.UUID.randomUUID().toString.take(5)}"
+
+      val leftSideRequiredSchema = getRequiredSchema(leftProjections, leftHolder)
+      val rightSideRequiredSchema = getRequiredSchema(rightProjections, rightHolder)
+
+      // Alias the duplicated columns from left side of the join.
+      val leftSideRequiredSchemaWithAliases = leftSideRequiredSchema.fields.map { field =>
+        val aliasName = if (rightSideRequiredSchema.fieldNames.contains(field.name)) {
+          generateJoinOutputAlias(field.name)
+        } else {
+          null
+        }
+
+        new SupportsPushDownJoin.ColumnWithAlias(field.name, aliasName)
+      }
+
+      // Aliasing of duplicated columns in right side of the join is not needed because the
+      // the conflicts are resolved by aliasing the left side.
+      val rightSideRequiredSchemaWithAliases = rightSideRequiredSchema.fields.map { field =>
+        new SupportsPushDownJoin.ColumnWithAlias(field.name, null)
+      }
+
+      // Create the AttributeMap that holds (Attribute -> Attribute with up to date name) mapping.
+      val pushedJoinOutputMap = AttributeMap[Expression](
+        node.output.asInstanceOf[Seq[AttributeReference]]
+          .zip(leftSideRequiredSchemaWithAliases ++ rightSideRequiredSchemaWithAliases)
+          .map{ case (attr, columnWithAlias) =>
+            if (columnWithAlias.getAlias == null) {
+              (attr, attr)
```

Comment:
When we rewrite attributes, we do `pushedJoinOutputMap.getOrElse(a, a)`. I think we don't need to put these entries into the map, as the key and value are the same.